core/corehttp/gateway_handler_unixfs__redirects.go

```diff
@@ -0,0 +1,289 @@
+			toPath := rule.To
+			toPath = replacePlaceholders(toPath, match)
+			toPath = replaceSplat(toPath, match)
+
+			// Rewrite
+			if rule.Status == 200 {
+				// Prepend the rootPath
+				toPath = rootPath + toPath
+				return false, toPath, nil
+			}
+
+			// Or 404
+			if rule.Status == 404 {
+				toPath = rootPath + toPath
+				content404Path := ipath.New(toPath)
+				err := i.serve404(w, r, content404Path)
+				return true, toPath, err
+			}
+
+			// Or redirect
+			http.Redirect(w, r, toPath, rule.Status)
+			return true, toPath, nil
+		}
+	}
+
+	// No redirects matched
+	return false, "", nil
+}
+
+func replacePlaceholders(to string, match urlpath.Match) string {
+	if len(match.Params) > 0 {
+		for key, value := range match.Params {
+			to = strings.ReplaceAll(to, ":"+key, value)
+		}
+	}
+
+	return to
+}
+
+func replaceSplat(to string, match urlpath.Match) string {
+	return strings.ReplaceAll(to, ":splat", match.Trailing)
+}
+
+func (i *gatewayHandler) getRedirectRules(r *http.Request, redirectsFilePath ipath.Resolved) ([]redirects.Rule, error) {
+	// Convert the path into a file node
+	node, err := i.api.Unixfs().Get(r.Context(), redirectsFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not get _redirects node: %v", err)
+	}
+	defer node.Close()
+
+	// Convert the node into a file
+	f, ok := node.(files.File)
+	if !ok {
+		return nil, fmt.Errorf("could not convert _redirects node to file")
+	}
+
+	// Parse redirect rules from file
+	redirectRules, err := redirects.Parse(f)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse redirect rules: %v", err)
+	}
+
+	return redirectRules, nil
+}
+
+// Returns a resolved path to the _redirects file located in the root CID path of the requested path
+func (i *gatewayHandler) getRedirectsFile(r *http.Request, logger *zap.SugaredLogger) ipath.Resolved {
+	// r.URL.Path is the full ipfs path to the requested resource,
+	// regardless of whether path or subdomain resolution is used.
+	rootPath, err := getRootPath(r.URL.Path)
+	if err != nil {
+		logger.Debugf("getRootPath failed", err)
+		return nil
+	}
+
+	// Check for root path.
+	_, err = i.api.Block().Get(r.Context(), rootPath)
+	if err != nil {
+		return nil
+	}
+
+	// Check for _redirects file.
+	// Any path resolution failures are ignored and we just assume there's no _redirects file.
+	// Note that ignoring these errors also ensures that the use of the empty CID (bafkqaaa) in tests doesn't fail.
+	path := ipath.Join(rootPath, "_redirects")
+	resolvedPath, err := i.api.ResolvePath(r.Context(), path)
+	if err != nil {
+		return nil
+	}
+	return resolvedPath
+}
+
+// Returns the root CID Path for the given path
+func getRootPath(path string) (ipath.Path, error) {
+	if isIpfsPath(path) {
+		parts := strings.Split(path, "/")
+		return ipath.New(gopath.Join(ipfsPathPrefix, parts[2])), nil
+	}
+
+	if isIpnsPath(path) {
+		parts := strings.Split(path, "/")
+		return ipath.New(gopath.Join(ipnsPathPrefix, parts[2])), nil
+	}
```

Comment:
nit: if you switch `path` from `string` to `contentPath ipath.Path`, then you will be able to read namespace via `contentPath.Namespace()` and simplify this